File: infra/scripts/yimba_scraper.js
/**
 * YIMBA.sk Real Estate Portal Scraper
 *
 * This scraper extracts real estate development project data from YIMBA.sk
 * YIMBA is an architectural/urban development information portal in Slovakia
 *
 * Key Findings:
 * - Backend: Fat-Free Framework (PHP)
 * - CDN: Cloudflare
 * - Data Format: JSON API available
 * - No authentication required
 * - No rate limiting observed
 * - robots.txt allows scraping
 */

const https = require('https');
const { chromium } = require('playwright'); // Optional: for detail pages

class YimbaScraper {
    constructor() {
        this.baseUrl = 'https://www.yimba.sk';
        this.headers = {
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
            'Accept': 'application/json, text/html, */*',
            'X-Requested-With': 'XMLHttpRequest',
            'Referer': 'https://www.yimba.sk/'
        };
    }

    /**
     * Get all projects with filtering options
     *
     * @param {Object} options - Filter options
     * @param {string|string[]} options.status - Project status: 1=Planning, 2=Under Construction, 3=Completed, 4=Cancelled
     * @param {string} options.sort - Sort field: 'name', 'date', etc.
     * @param {string} options.order - Sort order: 'asc' or 'desc'
     * @param {boolean} options.archive - Include archived projects (0 or 1)
     * @returns {Promise<Array>} Array of project objects
     */
    async getProjects(options = {}) {
        const {
            status = [1, 2, 3, 4], // All statuses by default
            sort = 'name',
            order = 'asc',
            archive = 0
        } = options;

        const statusParam = Array.isArray(status) ? status.join(',') : status;
        const url = `${this.baseUrl}/projekty/status/${statusParam}/zoradenie/${sort}/${order}/archive/${archive}?format=json`;

        console.log(`Fetching projects: ${url}`);

        return new Promise((resolve, reject) => {
            https.get(url, { headers: this.headers }, (res) => {
                let data = '';

                res.on('data', (chunk) => {
                    data += chunk;
                });

                res.on('end', () => {
                    try {
                        const projects = JSON.parse(data);
                        console.log(`Retrieved ${projects.length} projects`);
                        resolve(projects);
                    } catch (error) {
                        reject(new Error(`Failed to parse JSON: ${error.message}`));
                    }
                });
            }).on('error', reject);
        });
    }

    /**
     * Get projects by specific status
     * Status codes:
     * 1 = Planning/Proposed
     * 2 = Under Construction
     * 3 = Completed
     * 4 = Cancelled/Stopped
     */
    async getProjectsByStatus(status) {
        return this.getProjects({ status });
    }

    /**
     * Get project detail page HTML
     * Note: Detail data is embedded in HTML, not available as JSON API
     *
     * @param {string} slug - Project slug (e.g., 'ahoj-park')
     * @returns {Promise<string>} HTML content
     */
    async getProjectDetail(slug) {
        const url = `${this.baseUrl}/${slug}`;

        return new Promise((resolve, reject) => {
            https.get(url, { headers: this.headers }, (res) => {
                let data = '';

                res.on('data', (chunk) => {
                    data += chunk;
                });

                res.on('end', () => {
                    resolve(data);
                });
            }).on('error', reject);
        });
    }

    /**
     * Get project detail using headless browser (for dynamic content)
     * This is more resource-intensive but captures all data including lazy-loaded content
     *
     * @param {string} slug - Project slug
     * @returns {Promise<Object>} Extracted project data
     */
    async getProjectDetailDynamic(slug) {
        const browser = await chromium.launch({ headless: true });
        const page = await browser.newPage();

        try {
            await page.goto(`${this.baseUrl}/${slug}`, {
                waitUntil: 'domcontentloaded',
                timeout: 30000
            });

            // Extract data from the page
            const data = await page.evaluate(() => {
                const title = document.querySelector('h1')?.textContent?.trim();
                const description = document.querySelector('.project-description, .description')?.textContent?.trim();

                // Extract images
                const images = Array.from(document.querySelectorAll('img[src*="/upload/"]'))
                    .map(img => img.src);

                // Extract metadata
                const metaElements = document.querySelectorAll('.project-meta, .meta-item');
                const metadata = {};
                metaElements.forEach(el => {
                    const label = el.querySelector('.label, dt')?.textContent?.trim();
                    const value = el.querySelector('.value, dd')?.textContent?.trim();
                    if (label && value) {
                        metadata[label] = value;
                    }
                });

                return {
                    title,
                    description,
                    images,
                    metadata,
                    url: window.location.href
                };
            });

            await browser.close();
            return data;
        } catch (error) {
            await browser.close();
            throw error;
        }
    }

    /**
     * Search projects by keyword
     * Note: YIMBA doesn't expose a JSON search API, so this searches through all projects
     *
     * @param {string} keyword - Search keyword
     * @returns {Promise<Array>} Matching projects
     */
    async searchProjects(keyword) {
        const allProjects = await this.getProjects();
        const lowerKeyword = keyword.toLowerCase();

        return allProjects.filter(project =>
            project.name.toLowerCase().includes(lowerKeyword) ||
            (project.slug && project.slug.toLowerCase().includes(lowerKeyword))
        );
    }

    /**
     * Get statistics about projects
     */
    async getStatistics() {
        const allProjects = await this.getProjects();

        const stats = {
            total: allProjects.length,
            byStatus: {
                planning: 0,
                construction: 0,
                completed: 0,
                cancelled: 0
            }
        };

        allProjects.forEach(project => {
            const status = parseInt(project.status);
            switch (status) {
                case 1: stats.byStatus.planning++; break;
                case 2: stats.byStatus.construction++; break;
                case 3: stats.byStatus.completed++; break;
                case 4: stats.byStatus.cancelled++; break;
            }
        });

        return stats;
    }

    /**
     * Helper method to introduce delay between requests (good practice)
     */
    async delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// Example usage
async function main() {
    const scraper = new YimbaScraper();

    try {
        // 1. Get all projects
        console.log('\n=== Fetching all projects ===');
        const allProjects = await scraper.getProjects();
        console.log(`Total projects: ${allProjects.length}`);
        console.log('First project:', allProjects[0]);

        // 2. Get only projects under construction
        console.log('\n=== Projects under construction ===');
        const underConstruction = await scraper.getProjectsByStatus(2);
        console.log(`Under construction: ${underConstruction.length}`);

        // 3. Search for specific projects
        console.log('\n=== Searching for apartments (byty) ===');
        const apartments = await scraper.searchProjects('byt');
        console.log(`Found ${apartments.length} apartment projects`);
        if (apartments.length > 0) {
            console.log('Example:', apartments[0]);
        }

        // 4. Get statistics
        console.log('\n=== Project Statistics ===');
        const stats = await scraper.getStatistics();
        console.log(stats);

        // 5. Get project detail (HTML)
        console.log('\n=== Fetching project detail ===');
        const detailHtml = await scraper.getProjectDetail('ahoj-park');
        console.log(`Detail HTML length: ${detailHtml.length} characters`);

        // Uncomment to use dynamic scraping with Playwright
        // const detailData = await scraper.getProjectDetailDynamic('ahoj-park');
        // console.log('Detail data:', detailData);

    } catch (error) {
        console.error('Error:', error.message);
    }
}

// Run if executed directly
if (require.main === module) {
    main();
}

module.exports = YimbaScraper;

File: infra/scripts/yimba_scraper.py
#!/usr/bin/env python3
"""
YIMBA.sk Scraper - Python Implementation

Complete scraper for YIMBA.sk real estate development portal.
Supports both API-based listing scraping and HTML detail page parsing.

Usage:
    python yimba_scraper.py

Requirements:
    pip install requests beautifulsoup4
"""

import requests
import time
import json
from typing import List, Dict, Optional
from bs4 import BeautifulSoup


class YimbaScraper:
    """Scraper for YIMBA.sk real estate portal"""

    BASE_URL = "https://www.yimba.sk"

    def __init__(self, delay: float = 0.5):
        """
        Initialize the scraper

        Args:
            delay: Delay between requests in seconds (default 0.5)
        """
        self.delay = delay
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
            'Accept': 'application/json, text/html, */*',
            'X-Requested-With': 'XMLHttpRequest',
            'Referer': 'https://www.yimba.sk/'
        })

    def get_projects(self,
                     status: str = "1,2,3,4",
                     sort: str = "name",
                     order: str = "asc",
                     archive: int = 0) -> List[Dict]:
        """
        Get projects from the listing API

        Args:
            status: Comma-separated status IDs (1=Planning, 2=Construction, 3=Completed, 4=Cancelled)
            sort: Sort field (name, date)
            order: Sort order (asc, desc)
            archive: Include archived projects (0 or 1)

        Returns:
            List of project dictionaries
        """
        url = f"{self.BASE_URL}/projekty/status/{status}/zoradenie/{sort}/{order}/archive/{archive}"
        params = {"format": "json"}

        print(f"Fetching projects from: {url}")

        response = self.session.get(url, params=params, timeout=30)
        response.raise_for_status()

        projects = response.json()
        print(f"Retrieved {len(projects)} projects")

        return projects

    def get_projects_by_status(self, status: int) -> List[Dict]:
        """
        Get projects filtered by specific status

        Args:
            status: Status ID (1=Planning, 2=Construction, 3=Completed, 4=Cancelled)

        Returns:
            List of project dictionaries
        """
        return self.get_projects(status=str(status))

    def get_project_detail_html(self, slug: str) -> str:
        """
        Get project detail page HTML

        Args:
            slug: Project slug (e.g., 'ahoj-park')

        Returns:
            HTML content as string
        """
        url = f"{self.BASE_URL}/{slug}"
        print(f"Fetching detail page: {url}")

        response = self.session.get(url, timeout=30)
        response.raise_for_status()

        return response.text

    def parse_project_detail(self, html: str) -> Dict:
        """
        Parse project detail page HTML to extract structured data

        Args:
            html: HTML content from detail page

        Returns:
            Dictionary with extracted project data
        """
        soup = BeautifulSoup(html, 'html.parser')

        # Extract basic info
        title = soup.find('h1')
        title_text = title.get_text(strip=True) if title else None

        # Extract description
        description = soup.find('div', class_=['project-description', 'description', 'content'])
        description_text = description.get_text(strip=True) if description else None

        # Extract images
        images = []
        for img in soup.find_all('img'):
            src = img.get('src', '')
            if '/upload/' in src:
                # Make absolute URL if relative
                if not src.startswith('http'):
                    src = self.BASE_URL + src
                images.append(src)

        # Extract metadata (flexible approach)
        metadata = {}
        meta_containers = soup.find_all(['div', 'dl'], class_=['project-meta', 'meta', 'info'])
        for container in meta_containers:
            # Try dt/dd pattern
            dts = container.find_all('dt')
            dds = container.find_all('dd')
            for dt, dd in zip(dts, dds):
                key = dt.get_text(strip=True)
                value = dd.get_text(strip=True)
                metadata[key] = value

            # Try label/value pattern
            labels = container.find_all(class_=['label', 'key'])
            values = container.find_all(class_=['value', 'val'])
            for label, value in zip(labels, values):
                key = label.get_text(strip=True)
                val = value.get_text(strip=True)
                metadata[key] = val

        return {
            'title': title_text,
            'description': description_text,
            'images': images,
            'metadata': metadata
        }

    def get_project_detail(self, slug: str) -> Dict:
        """
        Get complete project detail (HTML + parsed data)

        Args:
            slug: Project slug

        Returns:
            Dictionary with parsed project data
        """
        html = self.get_project_detail_html(slug)
        data = self.parse_project_detail(html)
        data['slug'] = slug
        return data

    def search_projects(self, keyword: str) -> List[Dict]:
        """
        Search projects by keyword (searches through all projects locally)

        Args:
            keyword: Search term

        Returns:
            List of matching projects
        """
        all_projects = self.get_projects()
        keyword_lower = keyword.lower()

        matching = [
            project for project in all_projects
            if keyword_lower in project['name'].lower() or
               keyword_lower in project.get('slug', '').lower()
        ]

        print(f"Found {len(matching)} projects matching '{keyword}'")
        return matching

    def get_statistics(self) -> Dict:
        """
        Get statistics about all projects

        Returns:
            Dictionary with project statistics
        """
        all_projects = self.get_projects()

        stats = {
            'total': len(all_projects),
            'by_status': {
                'planning': 0,
                'construction': 0,
                'completed': 0,
                'cancelled': 0
            }
        }

        for project in all_projects:
            status = int(project['status'])
            if status == 1:
                stats['by_status']['planning'] += 1
            elif status == 2:
                stats['by_status']['construction'] += 1
            elif status == 3:
                stats['by_status']['completed'] += 1
            elif status == 4:
                stats['by_status']['cancelled'] += 1

        return stats

    def scrape_all_details(self, status: Optional[str] = None) -> List[Dict]:
        """
        Scrape details for all projects (or filtered by status)

        Args:
            status: Optional status filter (e.g., "2" for under construction)

        Returns:
            List of projects with full details
        """
        if status:
            projects = self.get_projects(status=status)
        else:
            projects = self.get_projects()

        results = []
        total = len(projects)

        for i, project in enumerate(projects, 1):
            print(f"\nScraping {i}/{total}: {project['name']}")

            try:
                # Merge list data with detail data
                detail = self.get_project_detail(project['slug'])
                full_data = {**project, **detail}
                results.append(full_data)

                # Rate limiting
                time.sleep(self.delay)

            except Exception as e:
                print(f"Error scraping {project['slug']}: {e}")
                results.append(project)  # Add without details

        return results

    def save_to_json(self, data: List[Dict], filename: str):
        """Save data to JSON file"""
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
        print(f"Saved {len(data)} projects to {filename}")


def main():
    """Example usage of the scraper"""

    scraper = YimbaScraper(delay=0.5)

    # 1. Get all projects
    print("\n=== Fetching all projects ===")
    all_projects = scraper.get_projects()
    print(f"Total projects: {len(all_projects)}")
    if all_projects:
        print(f"First project: {all_projects[0]}")

    # 2. Get projects under construction
    print("\n=== Projects under construction ===")
    under_construction = scraper.get_projects_by_status(2)
    print(f"Under construction: {len(under_construction)}")

    # 3. Get completed projects
    print("\n=== Completed projects ===")
    completed = scraper.get_projects_by_status(3)
    print(f"Completed: {len(completed)}")

    # 4. Search for apartments
    print("\n=== Searching for apartments ===")
    apartments = scraper.search_projects('byt')
    if apartments:
        print(f"Found {len(apartments)} apartment projects")
        print(f"Example: {apartments[0]['name']}")

    # 5. Get statistics
    print("\n=== Project Statistics ===")
    stats = scraper.get_statistics()
    print(json.dumps(stats, indent=2))

    # 6. Get detail for one project
    print("\n=== Fetching project detail ===")
    if all_projects:
        first_project = all_projects[0]
        detail = scraper.get_project_detail(first_project['slug'])
        print(f"Title: {detail.get('title')}")
        print(f"Images: {len(detail.get('images', []))}")
        print(f"Metadata keys: {list(detail.get('metadata', {}).keys())}")

    # 7. Optional: Scrape all details (commented out - takes time)
    # print("\n=== Scraping all construction projects with details ===")
    # full_data = scraper.scrape_all_details(status="2")
    # scraper.save_to_json(full_data, 'yimba_construction_projects.json')


if __name__ == '__main__':
    main()
